### ISSUE TITLE
CC0124 Replace if statement or operator with call to Math.Max

### ISSUE THREAD

**Issue by kwhitefoot** [code-cracker/code-cracker#799]
#New analyzer:

Replace if statement or operator with call to Math.Max. 

Max is easier to read and quite often shorter.

Before:

````vb.net
     a = If(b >= c, b, c) 

````

After:

````vb.net
  a = Math.Max(b, c)
````

You can add more information here, e.g. conditions under which a diagnostic should not trigger, etc.

Diagnostic Id: `CC0124`
Category: `Maintainability`
Severity: `Info`

**Comment by wmjordan** [comment]
Would this harm performance?
Calling the `Math.Max` could add extra stack pushes and pops. In performance critical scenario, the _"if then else"_ pattern might be preferred.

**Comment by kwhitefoot** [comment]
@wmjordan Good point.  On the other hand Math.Max more clearly expresses what the code _means_.
I suppose I should try to create a simple benchmark.

**Comment by kwhitefoot** [comment]
@wmjordan Here are the results of a simple benchmark.

Times in hours, minutes, seconds, and fractions of a second (Win7 64bit, Intel i7-4800MQ at 2.7GHz, 32GB RAM):

    simple     if: 00:00:00.0330000
    simple    max: 00:00:00.0920000
    work naive if: 00:00:42.7580000
    work       if: 00:00:28.6360000
    work      max: 00:00:28.6770000

So my conclusion is that max is not noticeably slower in many real world cases where:

- a and b are the results of functions and 
- you only need the maximum value and are not going to use a and b for anything else

Hard to tell for at a glance for any given case which will be faster.
My preference is for readability first, speed afterwards if necessary.  

But notice that to make the longest running version last 40 seconds I needed 10 million loops so unless this is executed a lot or is extremely time critical it probably isn't going to slow your program down noticeably whichever you choose.

<pre> 

  Sub Main()

    Const loops = 10000000

    Dim a = 1.234
    Dim b = 2.345

    Dim start, finish As DateTime
    start = Now
    For i = 1 To loops
      Dim c = If(a > b, a, b)
    Next
    finish = Now
    Console.WriteLine($"simple     if: {(finish - start)}")

    start = Now
    For i = 1 To loops
      Dim c = Math.Max(a, b)
    Next
    finish = Now
    Console.WriteLine($"simple    max: {(finish - start)}")

    start = Now
    For i = 1 To loops
      Dim c = If(DoALotOfWork(a) > DoALotOfWork(b), DoALotOfWork(a), DoALotOfWork(b))
    Next
    finish = Now
    Console.WriteLine($"work naive if: {(finish - start)}")

    start = Now
    For i = 1 To loops
      Dim aa = DoALotOfWork(a)
      Dim bb = DoALotOfWork(b)
      Dim c = If(aa > bb, aa, bb)
    Next
    finish = Now
    Console.WriteLine($"work       if: {(finish - start)}")

    start = Now
    For i = 1 To loops
      Dim c = Math.Max(DoALotOfWork(a), DoALotOfWork(b))
    Next
    finish = Now
    Console.WriteLine($"work      max: {(finish - start)}")

    Console.ReadLine()
  End Sub

  Function DoALotOfWork(a As Double) As Double
    Dim n = 0.0
    For i = 1 To 100
      n *= a
      n += a
      n /= a
    Next
    Return a
  End Function
</pre>

**Comment by wmjordan** [comment]
The _simple_ benchmark has indicated that `Max` is about 3 times slower than `if`.
_work naive if_ was meaningless since it has called the same function twice.

Since the method runs very quickly, performance difference could be unnoticeable. It does save some code lines in the _work_ scenario and your point does stand.

**Comment by kwhitefoot** [comment]
> work naive if was meaningless since it has called the same function twice.
Do you mean to say that you have never seen that in the wild?  Lucky you, I see code worse than that everyday, that's why I called it naive.

Kevin.

------
Unix is like a wigwam: no gates, no windows and an Apache inside.

Date: Thu, 2 Jun 2016 03:44:43 -0700

**Comment by giggio** [comment]
I don't like. I find it too specific, quite rare, and I don't think it adds much value.

**Comment by giggio** [comment]
I don't think this is on category `Maintainability`. Checkout some examples for this category here: https://msdn.microsoft.com/en-us/library/ms182211.aspx
